test/hotspot/jtreg/compiler/c2/irTests/TestScopedValue.java

```diff
@@ -0,0 +1,708 @@
+/*
+ * Copyright (c) 2024, Red Hat, Inc. All rights reserved.
```

Comment:
I like the tests, there is a lot of material here.

A few more ideas:
- have two scoped values, and then have a sequence of `get` and `getValue` calls on them, in some random mix. And check that everything gets commoned, and the result is correct.
- have a method that directly uses `get`, but also has inner scopes of `where`/`get`. Interleave these, maybe even with multiple different scoped values. And nest them with various depths. And then verify both the expected number of calls / loads, as well as the result.

Also: is it possible to stuff ScopedValues into ScopedValues? That would be another interesting stress-test with lots of options.

Reply:
In the commit that I will push soon, I added more tests: a couple with 3 scoped values and a few with ScopedValues into ScopedValues. The ones you suggest with has inner scopes of `where/get ` can't work because `Cache.invalidate()` would then be called: when C2 sees a call to `Cache.invalidate()` , it doesn't perform any optimization.